app/upgrades.go

```diff
@@ -1,3 +1,158 @@
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	distrkeeper "github.com/cosmos/cosmos-sdk/x/distribution/keeper"
+	distrtypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	communitytypes "github.com/kava-labs/kava/x/community/types"
+)
+
+const (
+	MainnetUpgradeName = "v0.22.0"
+	TestnetUpgradeName = "v0.22.0-alpha.0"
+)
+
+func (app App) RegisterUpgradeHandlers() {
+	// register upgrade handler for mainnet
+	app.upgradeKeeper.SetUpgradeHandler(MainnetUpgradeName,
+		func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+			app.Logger().Info("running mainnet upgrade handler")
+
+			toVM, err := app.mm.RunMigrations(ctx, app.configurator, fromVM)
+			if err != nil {
+				return toVM, err
+			}
+
+			app.Logger().Info("move all community pool funds from x/distribution to x/community")
+			FundCommunityPoolModule(ctx, app.distrKeeper, app.bankKeeper, app)
+
+			app.Logger().Info("granting x/gov module account x/community module authz messages")
+			GrantGovCommunityPoolMessages(ctx, app.authzKeeper, app.accountKeeper)
+
+			return toVM, nil
+		},
+	)
+
+	// register upgrade handler for testnet
+	app.upgradeKeeper.SetUpgradeHandler(TestnetUpgradeName,
+		func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+			app.Logger().Info("running testnet upgrade handler")
+
+			toVM, err := app.mm.RunMigrations(ctx, app.configurator, fromVM)
+			if err != nil {
+				return toVM, err
+			}
+
+			app.Logger().Info("move all community pool funds from x/distribution to x/community")
+			FundCommunityPoolModule(ctx, app.distrKeeper, app.bankKeeper, app)
+
+			app.Logger().Info("granting x/gov module account x/community module authz messages")
+			GrantGovCommunityPoolMessages(ctx, app.authzKeeper, app.accountKeeper)
+
+			return toVM, nil
+		},
+	)
+
+	upgradeInfo, err := app.upgradeKeeper.ReadUpgradeInfoFromDisk()
+	if err != nil {
+		panic(err)
+	}
+
+	// note: no store updates
+	doUpgrade := upgradeInfo.Name == MainnetUpgradeName || upgradeInfo.Name == TestnetUpgradeName
+	if doUpgrade && !app.upgradeKeeper.IsSkipHeight(upgradeInfo.Height) {
+		storeUpgrades := storetypes.StoreUpgrades{}
+
+		// configure store loader that checks if version == upgradeHeight and applies store upgrades
+		app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, &storeUpgrades))
+	}
+}
+
+// GrantGovCommunityPoolMessages grants x/gov module account access to submit x/authz messages from the community pool module account.
+func GrantGovCommunityPoolMessages(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	accountKeeper authkeeper.AccountKeeper,
+) {
+	communityAddr := accountKeeper.GetModuleAddress(communitytypes.ModuleName)
+	govAddr := accountKeeper.GetModuleAddress(govtypes.ModuleName)
+	allowedMsgs := []string{
```

Comment:
i'd consider pulling this list out into a variable that could be imported by an e2e test that does acceptance to check that each message has actually been granted.